### ISSUE TITLE
delta transfer algorithm broken, since v3.2.4

### ISSUE THREAD

**Issue by Tunoac** [WayneD/rsync#317]
Since v3.2.4 the delta transfer algorithm is broken - on some version<>architecture combinations.
As a result big files are transferred whole again, not only the changed part.

used test file with 4MB size:
```
dd if=/dev/urandom of=foobar0  bs=1M count=4
cp foobar0 foobar1
date >> foobar1
```
--> so foobar0 and foobar 1 are almost the same, except the appended date.

used rsync command lines:
```
rsync  --inplace --no-whole-file --partial --progress --archive --human-readable --stats --checksum foobar0 server:foobar0
rsync  --inplace --no-whole-file --partial --progress --archive --human-readable --stats --checksum foobar1 server:foobar1
```
--> on 2nd command is expected to transfer only the small changed part, but instead sometimes the whole file is transferred, logfile with -vvvv captured.

test result regarding delta transfer:

```
aarch64_(v3.2.3)--to--armv7l_(v3.2.4)	fail
aarch64_(v3.2.3)--to--x86_64_(v3.2.4)	ok

aarch64_(v3.2.4)--to--armv7l_(v3.2.4)	ok
aarch64_(v3.2.4)--to--x86_64_(v3.2.4)	fail

x86_64_(v3.2.4)--to--armv7l_(v3.2.4)	fail
x86_64_(v3.2.4)--to--x86_64_(v3.2.4)	ok

```
Building a x86_64_(v3.2.4) version with --disable-roll-simd does not solve the issue.
I'm happy to provide more info if required, or performing another test-case
OS is archlinux and archlinuxarm for this test setup.

[aarch64_(v3.2.3)--to--armv7l_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645141/aarch64_.v3.2.3.--to--armv7l_.v3.2.4.txt)
[aarch64_(v3.2.3)--to--x86_64_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645142/aarch64_.v3.2.3.--to--x86_64_.v3.2.4.txt)
[aarch64_(v3.2.4)--to--armv7l_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645144/aarch64_.v3.2.4.--to--armv7l_.v3.2.4.txt)
[aarch64_(v3.2.4)--to--x86_64_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645145/aarch64_.v3.2.4.--to--x86_64_.v3.2.4.txt)
[x86_64_(v3.2.4)--to--armv7l_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645146/x86_64_.v3.2.4.--to--armv7l_.v3.2.4.txt)
[x86_64_(v3.2.4)--to--x86_64_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645147/x86_64_.v3.2.4.--to--x86_64_.v3.2.4.txt)
[aarch64_(v3.2.3).txt](https://github.com/WayneD/rsync/files/8645148/aarch64_.v3.2.3.txt)
[aarch64_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645149/aarch64_.v3.2.4.txt)
[x86_64_(v3.2.4).txt](https://github.com/WayneD/rsync/files/8645150/x86_64_.v3.2.4.txt)
.

**Comment by paulmenzel** [comment]
Could you bisect the commit introducing the issue?

**Comment by WayneD** [comment]
The important part of this is the list of generated checksums. It appears that the non-x86 versions are generating incorrect checksums while the others do not. The workaround should be to compile them with `--disable-roll-simd` not the x86 version.

**Comment by Tunoac** [comment]
I'm not very familiar with git and bisect, but I think I understood your request. I did several manual builts with the 4-2-1-0-rule, compiling on a raspberry pi3 (armv7l).

basis:
`git log --pretty=oneline v3.2.3 v3.2.4`

result:
```
bad	0ac7ebceef70417355f25daf9e2fd94e84c49749 Preparing for release of 3.2.4
	...
bad	9fc7deab0d917db5ddcdc752be13853039c8f877 Update CI builds to new path-setting idiom.
	b115bc8a5d5262ba06a573ef4f54eba806b1200a Silence a few more warnings.
	cd018c7a4c40a4d4f2aa3ea24f77282efe213c40 Use a better -Wno-pedantic heuristic.
	9fce0eb5ab80da4d4d97bcde346c9b0b92432ac8 Avoid some pedantic errors & old warnings.
	33e94849b1013ff7e7b04e0a45ab2fffe2dd1e44 Handle early gcc versions that don't understand -Wno-pedantic.
bad	8f1511184ab5ad10110286cdef22195b5a67f01f Make gcc die on init overflow of an array.
good	acca9d43d3143113088efa76d561b813735f1a73 Expand the max name_num_item list size.
good	58f464f4daa1b1ee00c6afd898a19ca5cd81bf58 Change `--info=skip2` messages & add info on attr changes.
good	7eb59a9152a2ace7bc7858e9915c671b3ab54344 Change from $build_cpu to $host_cpu as edo1 suggested.
	740ed11aa82bb61022d927e669e3c79f704c941b Make the extra info on the "exists" messages optional.
	d2a97a7ab492e0d0548708251309e077e6aa8c8b Various file comparison improvements
good	15bc7ded398147ee50232b73c804f10aafa2fdda More NEWS updates.
	...
good	e94bad1c156fc3910f24e2b3b71a81b0b0bdeb70 Preparing for release of 3.2.3
```

**Comment by paulmenzel** [comment]
Very nice. There is the command [`git bisect`](https://git-scm.com/docs/git-bisect) making this easy and faster by doing a binary search. But it looks like you find the culprit.

So your search says that commit https://github.com/WayneD/rsync/commit/8f1511184ab5ad10110286cdef22195b5a67f01f (Make gcc die on init overflow of an array.) causes the regression. On the master branch, if you do

    git revert 8f1511184ab5ad10110286cdef22195b5a67f01f

and fix the conflict in `configure.ac` by just deleting the tagged lines, does it work again?

**Comment by WayneD** [comment]
I got access to an armv8 host and confirmed the error. Then, I ran `make distclean` and removed the part in configure.ac that added the `-pedantic-errors` option (configure & make) and the resulting rsync then worked fine.  So, this looks to be a compiler issue.

**Comment by WayneD** [comment]
OK, I've committed a work-around that avoids that compiler option on non-x86 systems as a quick fix.  The latest git source should work fine on arm, and someone packaging 3.2.4 for arm could make the same change to the 3.2.4 configure.ac or just delete that chunk of shell code that adds the option.

Thanks for the help!

**Comment by ismail** [comment]
@WayneD Could you update https://download.samba.org/pub/rsync/src/rsync-patches-3.2.4.tar.gz with the patch? That would make it easy for Homebrew to apply the patch. Thanks!

**Comment by WayneD** [comment]
I don't want to modify an official download file, but I could make the patch available on download.samba.org.  However, I'm currently zeroing in on a more exact reason for the failure.  It appears that the addition of the compile option early in the configure run causes the check for SIGNED_CHAR_OK to fail, and that lack of "schar" being defined as "signed char" is the real reason for the checksum code to not work right.

**Comment by Tunoac** [comment]
Yes, latest git is working fine for me, thanks @WayneD

**Comment by WayneD** [comment]
OK, I have pushed out a better configure fix, one that fixes SIGNED_CHAR_OK for everyone.  Apparently arm defaults "char" to unsigned characters (unlike x86), which is why the issue was seen for arm systems but not x86.

**Comment by ismail** [comment]
The standard complaints way to check for signed char seems to be including limits.h and checking if CHAR_MIN is 0 (unsigned case):

```
#include <stdio.h>
#include <limits.h>

int main(void)
{
    printf("chars are %ssigned\n", CHAR_MIN == 0? "un": "");
}
```

You can confirm this with by compiling with `-fsigned-char` and `-funsigned-char`, here is a diff for configure.ac:

```
-AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[]], [[signed char *s = ""]])],[rsync_cv_SIGNED_CHAR_OK=yes],[rsync_cv_SIGNED_CHAR_OK=no])])
+AC_RUN_IFELSE([AC_LANG_SOURCE([[
+#include <limits.h>
+int main() {
+    if (CHAR_MIN == 0)
+        return 0;
+    else
+        return 1;
+} ]])],[rsync_cv_SIGNED_CHAR_OK=yes],[rsync_cv_SIGNED_CHAR_OK=no])])
```

**Comment by WayneD** [comment]
We're actually checking if "signed char" is accepted as an explicit request, not what the default is.

**Comment by WayneD** [comment]
There is now a patch file available for this issue:

https://download.samba.org/pub/rsync/rsync-schar-fix-3.2.4.diff

**Comment by fxcoudert** [comment]
The check for `signed char` (which is an ANSI C type) was introduced in 1998 (https://github.com/WayneD/rsync/commit/debb45054a3c4d4da3faac372e993b594e5a0df7) because:

> Solaris 2.6 doesn't support "signed char". What a broken system!

In 2022 it might not be useful anymore to keep such checks.

**Comment by WayneD** [comment]
This fix is going out in 3.2.5.
